### ISSUE TITLE
p:www-form-urldecode: relax restriction on names of www-form-urlencoded parameters

### ISSUE THREAD

**Issue by Conal-Tuohy** [xproc/3.0-steps#22]
The `p:www-form-urldecode` step provides a way to decode submissions produced by HTML forms, but because only names that match the `xs:NCName` pattern are allowed, the step can't be used to decode certain form submissions which are themselves perfectly valid as HTTP POST payloads or as the query component of a URI. 

For example, the step could not parse submissions from the following form, because the `input`'s name contains a space:
```html
<form action="/my-form-handling-xproc-pipeline" method="get">
   <div>
      <input type="text" name="example input"/>
      <button type="submit">submit example</button>
   </div>
</form>
```
Submitting this form would produce the request URI `/my-form-handling-xproc-pipeline?example+input=` which `p:www-form-urldecode` would be unable to parse.

This leaves XProc developers without a convenient method to handle arbitrary forms or URIs with arbitrary query components.

The inverse step, `p:www-form-urlencode`, is also unnecessarily restricted, but this in practical terms this restriction is less of a obstacle since it's easier to perform the encoding than the decoding.

The specification for the `p:www-form-urldecode` step says:

> It is a dynamic error (err:XC0061) if the name of any encoded parameter name is not a valid `xs:NCName`. In other words, this step can only decode simple name/value pairs where the names do not contain colons or any characters that cannot be used in XML names.

However, the parameter names submitted from an HTML form are the values of `name` attributes of controls on that form, which are defined as `CDATA`; a much broader space than `xs:NCName`. 

Although a `c:param-set` element which contains a `c:param` whose name is not a valid `xs:NCName` will not be usable as the input to a `parameters` port, this requirement is not a good reason to impose the `xs:NCName` restriction on the output of the `p:www-form-urldecode` step. If a pipeline author wants to use parameters produced by `p:www-form-urldecode` as the input to a `parameter` port, they can always filter the parameters first. Invalidly-named parameters should only cause an exception when they are used as input to a `parameter`-type port.

**Comment by ndw** [comment]
Further consensus on 10 June 2019; we'll leave them steps for consistency with the rest of the design in 3.0; encoding will take a map and return a string; decoding will take a string and return a map.

**Comment by Conal-Tuohy** [comment]
I presume www-form-urldecode will produce a JSON document consisting of a map of strings to arrays of strings? i.e. map(xs:string, array(xs:string))

**Comment by xml-project** [comment]
I do not understand why the result should be "a map of strings to arrays of strings". Where does the array on the map's value result from, i.e. which rule should be used to split the string into pieces?

**Comment by xml-project** [comment]
I think we should use `map(xs:string, xs:string+)` for duplicate keys, so we are in line with the behaviour of map:merge() with `duplicates='combine'`.

**Comment by xml-project** [comment]
fixed with pr #147
